### PR TITLE
Add support for typo tolerance customization

### LINKED ISSUE #260: Add support to the typo tolerance customization
⚠️ This issue is generated, it means the examples and the namings do not necessarily correspond to the language of this repository.
Also, if you are a maintainer, feel free to add any clarification and instruction about this issue.

Related to:
- parent issue: https://github.com/meilisearch/integration-guides/issues/190
- specification: https://github.com/meilisearch/specifications/pull/117

----

Add a `typoTolerance` index settings resource to manage customization of the typo tolerance feature at the index level.

- Expose `typoTolerance` object on `/indexes/:index_uid/settings` endpoints.
- Add `GET/POST/DELETE` - `/indexes/:index_uid/settings/typo-tolerance` endpoints.

The `typoTolerance` object definition is made of the following properties:
  - `enabled` type (`boolean`)
  - `disableOnAttributes` type (`array[string]`)
  - `disableOnWords` type (`array[string]`)
  - `minWordSizeForTypos` type (`object`)
    - `oneTypo` type (`int`)
    - `twoTypos` type (`int`)

More information about what each field does could be [checked in the spec.](https://github.com/meilisearch/specifications/pull/117)

Check the other customization methods for reference like: `synonyms`, `searchableAttributes`, `filterableAttributes`.

TODO:
- [ ] Add the new methods to customize `typo-tolerance`
  - [ ] GET `/indexes/:index_uid/settings/typo-tolerance`
  - [ ] POST `/indexes/:index_uid/settings/typo-tolerance`
  - [ ] DELETE `/indexes/:index_uid/settings/typo-tolerance`
- [ ] Add tests

### PR DESCRIPTION
**Pull request by vishalsodani**
# Pull Request

## Related issue
Fixes #260 

## What does this PR do?
- ...

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
Hey @vishalsodani thanks for this contribution :) 

Could you look at [the specification](https://github.com/meilisearch/specifications/blob/main/text/0117-typo-tolerance-setting-api.md#31-typotolerance-api-resource-definition)? I think the parameters are optional

**Comment by bidoubiwa** [comment]
I know it was not part of the PR description, but could you add (or update) the following code-samples based on [this example](https://github.com/meilisearch/documentation/blob/master/.code-samples.meilisearch.yaml):

- get_typo_tolerance_1
- update_typo_tolerance_1
- reset_typo_tolerance_1
- typo_tolerance_guide_1
- typo_tolerance_guide_2
- typo_tolerance_guide_3
- typo_tolerance_guide_4
- settings_guide_typo_tolerance_1
- getting_started_typo_tolerance
- update_settings_1

**Comment by vishalsodani** [comment]
> I know it was not part of the PR description, but could you add (or update) the following code-samples based on [this example](https://github.com/meilisearch/documentation/blob/master/.code-samples.meilisearch.yaml):
> 
>     * get_typo_tolerance_1
> 
>     * update_typo_tolerance_1
> 
>     * reset_typo_tolerance_1
> 
>     * typo_tolerance_guide_1
> 
>     * typo_tolerance_guide_2
> 
>     * typo_tolerance_guide_3
> 
>     * typo_tolerance_guide_4
> 
>     * settings_guide_typo_tolerance_1
> 
>     * getting_started_typo_tolerance
> 
>     * update_settings_1

ok. I will add these.

**Comment by vishalsodani** [comment]
> Hey @vishalsodani thanks for this contribution :)
> 
> Could you look at [the specification](https://github.com/meilisearch/specifications/blob/main/text/0117-typo-tolerance-setting-api.md#31-typotolerance-api-resource-definition)? I think the parameters are optional

Can you please review? I have made the parameters optional. Is the implementation ok?

File: src/settings.rs [hunk 1]
```diff
@@ -979,6 +1062,43 @@ impl Index {
...
+    /// # let MEILISEARCH_API_KEY = option_env!("MEILISEARCH_API_KEY").unwrap_or("masterKey");
+    /// #
+    /// # futures::executor::block_on(async move {
+    /// let client = Client::new(MEILISEARCH_HOST, MEILISEARCH_API_KEY);
+    /// # client.create_index("set_typo_tolerance", None).await.unwrap().wait_for_completion(&client, None, None).await.unwrap();
+    /// let mut index = client.index("set_typo_tolerance");
+    ///
+    /// let mut typo_tolerance = TypoToleranceSettings::default();
+    ///
+    /// let task = index.set_typo_tolerance(&typo_tolerance).await.unwrap();
```
Comment (bidoubiwa) [inline src/settings.rs:1082]: Could you add different values than the default ones?

**Review by bidoubiwa — COMMENTED** [review]
Very nice PR 🔥

**Comment by bidoubiwa** [comment]
Hey @vishalsodani `one_type` and `two_typos` are `u8` and not `i64`. [See here](https://github.com/meilisearch/meilisearch/blob/77cdecaf3a45132bae5359088de2e00756c01482/meilisearch-lib/src/index/updates.rs#L44)

**Comment by vishalsodani** [comment]
Hi,

I feel there is an issue with this API design. If I tun this test, the test fails

`

         #[meilisearch_test]

         async fn test_set_typo_tolerance(client: Client, index: Index) {

        let typo_tolerance = TypoToleranceSettings {
            enabled: None,
            disable_on_attributes: Some(vec![]),
            disable_on_words: Some(vec![]),
            min_word_size_for_typos: Some(MinWordSizeForTypos {
                one_typo: Some(6),
                two_typos: Some(9),
            }),
        };
        let task_info = index.set_typo_tolerance(&typo_tolerance).await.unwrap();
        client.wait_for_task(task_info, None, None).await.unwrap();

        let res = index.get_typo_tolerance().await.unwrap();

        assert_eq!(typo_tolerance, res);
    }
`
`

        left: `TypoToleranceSettings { enabled: None, disable_on_attributes: Some([]), disable_on_words: Some([]), min_word_size_for_typos: Some(MinWordSizeForTypos { one_typo: Some(6), two_typos: Some(9) }) }`,

       right: `TypoToleranceSettings { enabled: Some(true), disable_on_attributes: Some([]), disable_on_words: Some([]), min_word_size_for_typos: Some(MinWordSizeForTypos { one_typo: Some(6), two_typos: Some(9) }) }`', src/settings.rs:1609:9
We are setting `enabled` as None which should not be allowed. Am I correct?

**Comment by vishalsodani** [comment]
> I know it was not part of the PR description, but could you add (or update) the following code-samples based on [this example](https://github.com/meilisearch/documentation/blob/master/.code-samples.meilisearch.yaml):
> 
>     * get_typo_tolerance_1
> 
>     * update_typo_tolerance_1
> 
>     * reset_typo_tolerance_1
> 
>     * typo_tolerance_guide_1
> 
>     * typo_tolerance_guide_2
> 
>     * typo_tolerance_guide_3
> 
>     * typo_tolerance_guide_4
> 
>     * settings_guide_typo_tolerance_1
> 
>     * getting_started_typo_tolerance
> 
>     * update_settings_1

Added.

**Comment by vishalsodani** [comment]
> Hey @vishalsodani thanks for this contribution :)
> 
> Could you look at [the specification](https://github.com/meilisearch/specifications/blob/main/text/0117-typo-tolerance-setting-api.md#31-typotolerance-api-resource-definition)? I think the parameters are optional

I have redesigned the API to use a builder pattern. So, a user is able to create a default object and can then set various attributes.

**Comment by vishalsodani** [comment]
@bidoubiwa Hi, I am done with all the changes. Please review. Thanks!

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
Hey @vishalsodani 

> We are setting enabled as None which should not be allowed. Am I correct?

Actually, you are right and just raised a bug in Meilisearch haha! See [this spec](https://github.com/meilisearch/specifications/blob/main/text/0117-typo-tolerance-setting-api.md#32222-errors)

> 🔴 Sending a value with a different type than Boolean for the enabled field returns an [bad_request](https://github.com/meilisearch/specifications/blob/main/text/0061-error-format-and-definitions.md#bad_request) error.

While I appreciate a lot that you made so many checks 🔥, They checks should be done at Meilisearch side and not ours. As, if we start going that path, we are basically re-writing the error handler of Meilisearch.

Meanwhile, for us, `null` is accepted until fixed by Meilisearch in a next release. 

Can you remove: 
- default_as_true
- default_one_typo_size
- default_two_typos_size
- is_true
- is_default_for_one_typo
- is_default_for_two_typos
- is_min_word_size_object_default

**Comment by vishalsodani** [comment]
> * default_one_typo_size

ok. Should I also revert the builder pattern? It does not seem that useful now.

**Comment by vishalsodani** [comment]
> Hey @vishalsodani
 
> Actually, you are right and just raised a bug in Meilisearch haha! See [this spec](https://github.com/meilisearch/specifications/blob/main/text/0117-typo-tolerance-setting-api.md#32222-errors)

Great!

**Comment by vishalsodani** [comment]
@bidoubiwa 
> Can you remove:
Done

File: .code-samples.meilisearch.yaml [hunk 2]
```diff
@@ -152,6 +152,17 @@ update_settings_1: |-
...
   synonyms.insert(String::from("logan"), vec!["wolverine"]);
 
+  let min_word_size_for_typos = MinWordSizeForTypos {
+    one_typo: Some(4),
+    two_typos; Some(12)
+  }
+  let typo_tolerance = TypoToleranceSettings {
+    enabled: Some(true),
+    disable_on_attributes: Some(vec!["title".to_string()]),
+    disable_on_words: Some(vec![])
```
Comment (bidoubiwa) [inline .code-samples.meilisearch.yaml:162]:
```suggestion
```

The fields that are not required should not be added to the samples. Normally this should not raise an error, right?

File: src/settings.rs [hunk 3]
```diff
@@ -1428,4 +1579,50 @@ mod tests {
...
+
+        let task_info = index.set_typo_tolerance(&typo_tolerance).await.unwrap();
+        client.wait_for_task(task_info, None, None).await.unwrap();
+
+        let res = index.get_typo_tolerance().await.unwrap();
+
+        assert_eq!(typo_tolerance, res);
+    }
+
+    #[meilisearch_test]
```
Comment (bidoubiwa) [inline src/settings.rs:1609]: Can you add a test where you only send a partial object ? 
for example:

- update only `enabled`
- make the request, check if correct

```rust
 let typo_tolerance = TypoToleranceSettings {
    enabled: Some(false),
};
```

in the same test: 
- update only `disabled_on_attributes`
- make the request, check if both `enabled` and `disabled_on_attributes` still have the correct changes.

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
In your code-samples, could you only add the fields that are required and removed the ones that are not? Example in comment.

File: src/settings.rs (hunk 3 above)
Comment (vishalsodani) [reply src/settings.rs:1609]: I have to initialize all the fields of a struct. I cannot have a partial struct. Is it possible?

Comment (vishalsodani) [reply src/settings.rs:1609]: `

      missing structure fields:
      - disable_on_attributes
      -  disable_on_words
     - min_word_size_for_typos

`

Comment (bidoubiwa) [reply src/settings.rs:1609]: Oh yeah... I mixed up the languages. My bad. I guess there that's the reason for adding the builder pattern... If you don't mind can you add it? With all the default values set to None?

Very sorry about the mixup. A lot of context switches between languages😅

Comment (vishalsodani) [reply src/settings.rs:1609]: If you look at this sample code https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d78d2a9f68131c270eadffb4df44d4a0; the changes I had made allowed for partial objects to be sent. The standard output shows - 



      Ok("{\"minWordSizeForTypos\":{\"oneTypo\":0},\"age\":10}")

Comment (vishalsodani) [reply src/settings.rs:1609]: > If you look at this sample code https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d78d2a9f68131c270eadffb4df44d4a0; the changes I had made allowed for partial objects to be sent. The standard output shows -
> 
> ```
>   Ok("{\"minWordSizeForTypos\":{\"oneTypo\":0},\"age\":10}")
> ```

@bidoubiwa Please see the above sample code shared on playground.

**Comment by bidoubiwa** [comment]
_This message is sent automatically_

Thanks again for contributing to Meilisearch :heart:
If you are participating in [Hacktoberfest](https://blog.meilisearch.com/contribute-hacktoberfest-2022/), and you would like to receive some gift from Meilisearch too, please complete [this form](https://meilisearch.typeform.com/to/idt5qugY).

**Comment by bidoubiwa** [comment]
Hey @vishalsodani 
I'm a bit underwater right now. I gave you the `hacktoberfest-accepted` label until I have time to come back on this issue

**Comment by omid** [comment]
What's the state of this PR?
I can also help to improve it if there is something missing?

We need this.

File: src/settings.rs [hunk 4]
```diff
@@ -13,6 +13,43 @@ pub struct PaginationSetting {
...
+}
+
+impl Default for MinWordSizeForTypos {
+    fn default() -> Self {
+        MinWordSizeForTypos {
+            one_typo: Some(5),
+            two_typos: Some(9),
+        }
+    }
+}
```
Comment (irevoire) [inline src/settings.rs:30]: We shouldn't try to guess the default value of meilisearch by default, + we don't want to overwrite a previous call of the route.
Sending `None` is enough.

```suggestion
impl Default for MinWordSizeForTypos {
    fn default() -> Self {
        MinWordSizeForTypos {
            one_typo: None,
            two_typos: None,
        }
    }
}
```

File: src/settings.rs [hunk 5]
```diff
@@ -13,6 +13,43 @@ pub struct PaginationSetting {
...
+impl Default for TypoToleranceSettings {
+    fn default() -> Self {
+        TypoToleranceSettings {
+            enabled: Some(true),
+            disable_on_attributes: Some(vec![]),
+            disable_on_words: Some(vec![]),
+            min_word_size_for_typos: Some(MinWordSizeForTypos::default()),
+        }
+    }
+}
```
Comment (irevoire) [inline src/settings.rs:51]: Once again, no reason to force some settings when creating the default value. 

```suggestion
impl Default for TypoToleranceSettings {
    fn default() -> Self {
        TypoToleranceSettings {
            enabled: None,
            disable_on_attributes: None,
            disable_on_words: None,
            min_word_size_for_typos: None,
        }
    }
}
```

**Comment by omid** [comment]
@irevoire, the original author didn't have any activity on GitHub since then.
Seems like somebody else needs to take control.

**Comment by omid** [comment]
It's been merged via the following PR: https://github.com/meilisearch/meilisearch-rust/pull/480
@bidoubiwa let's close this also 🙏🏼

**Comment by bidoubiwa** [comment]
Absolutely :)

**Pull request closed without merging**